runtime/lua/vim/lsp/buf.lua

```diff
@@ -76,8 +76,74 @@ end
 ---     - on_list: (function) |lsp-on-list-handler| replacing the default handler.
 ---       Called for any non-empty result.
 function M.definition(options)
+  options = options or {}
   local params = util.make_position_params()
-  request_with_options(ms.textDocument_definition, params, options)
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@param results table<integer, {error: lsp.ResponseError, result: lsp.Location|lsp.Location[]|lsp.LocationLink[]|nil}>
+  local function handler(results)
```

Comment:
Hmm, the definition method now uses a separate handler; but I don't think such a duplicate is a good idea. Would it be possible to consolidate?

https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L450

Reply:
I agree that duplication is not optimal. As I mentioned in the PR description, ideally the same should be applied to `type_definition`, `declaration`, and `implementation` as they also can receive locations from multiple clients and all use `location_handler` currently. There are a couple of reasons why a custom handler was introduced:
- it merges and flattens potentially heterogeneous responses from multiple clients
- it has `options` in a closure to check for `on_list`
- since the responses come from multiple lsp's, it has to do extra work to use the correct client offset_encoding to turn results into location items via [`util.locations_to_items`](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L430) or [`util.jump_to_locations`](https://github.com/neovim/neovim/blob/master/runtime/lua/vim/lsp/handlers.lua#L438) in case there is only single result.

Alternatively, if you agree that other location handling lsp methods should be using the same mechanism. I can update the location_handler you linked to pretty much this implementation that would handle results from multiple clients. And update `vim.lsp.buf_request_all` to accept an extra parameter `options` that would be passed to handler along with `results`.

This way the `vim.lsp.buf.defintion` and the other location methods would looks like this

```lua
function M.definition(options)
  local params = util.make_position_params()
  local bufnr = api.nvim_get_current_buf()

  vim.lsp.buf_request_all(bufnr, ms.textDocument_definition, params, location_handler, options)
end
```

If we can can have this mechanism for other lsp methods that handle locations, the current location handler(the function you linked) will be no longer needed and could be removed.

What do you think?